extras/ams_wrapper/src/preprocessing/preprocess_image.py

```diff
@@ -0,0 +1,107 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+from typing import Tuple
+
+import tensorflow as tf
+import numpy as np
+
+
+class ImageDecodeError(ValueError):
+    pass
+
+
+class ImageResizeError(ValueError):
+    pass
+
+
+class ImagePreprocessError(ValueError):
+    pass
+
+
+def preprocess_binary_image(image: bytes, channels: int = None,
+                            target_size: Tuple[int, int] = None,
+                            channels_first=True,
+                            dtype=tf.dtypes.uint8, scale: float = None,
+                            standardization=False,
+                            reverse_input_channels=False) -> np.ndarray:
+    """
+    Preprocess binary image in PNG, JPG or BMP format, producing numpy array as a result.
+
+    :param image: Image bytes
+    :param channels: Number of image's channels
+    :param target_size: A tuple of desired height and width
+    :param channels_first: If set to True, image array will be in NCHW format,
+     NHWC format will be used otherwise
+    :param dtype: Data type that will be used for decoding
+    :param scale: If passed, decoded image array will be multiplied by this value
+    :param standardization: If set to true, image array values will be standarized
+    to have mean 0 and standard deviation of 1
+    :param reverse_input_channels: If set to True, image channels will be reversed
+    from RGB to BGR format
+    :raises ImageDecodeError(ValueError): if image cannot be decoded
+    :raises ImageResizeError(ValueError): if image cannot be resized
+    :raises ImagePreprocessError(ValueError): if image cannot be preprocessed
+    :returns: Preprocessed image as numpy array
+    """
+
+    try:
+        decoded_image = tf.io.decode_image(image, channels=channels, dtype=dtype)
+    except Exception as e:
+        raise ImageDecodeError('Provided image is invalid, unable to decode.') from e
+
+    if target_size:
+        try:
+            height, width = target_size
+            decoded_image = tf.image.resize_with_crop_or_pad(decoded_image, height, width)
+        except Exception as e:
+            raise ImageResizeError('Failed to resize provided binary image from: {} '
+                                   'to: {}.'.format(tf.shape(decoded_image), target_size)) from e
+
+    try:
+        image_array = decoded_image.numpy()
+        if standardization:
+            decoded_image = tf.image.per_image_standardization(decoded_image)
+        if reverse_input_channels:
+            image_array = image_array[..., ::-1]
+        if channels_first:
+            image_array = np.transpose(image_array, [2, 0, 1])
+        if scale:
+            image_array = image_array * scale
+    except Exception as e:
+        raise ImagePreprocessError('Failed to preprocess binary image, '
```

Comment:
I'm not sure about this error name. Above we have ImageResizeError. If we consider resizing as part of preprocessing then maybe this error name could be more specific. 

Also the part _"Failed to preprocess **binary** image..."_ in error message indicates working with binary format of the image while we start this `try...except...` block with `image_array = decoded_image.numpy()` and `decoded_image` is `Tensor` object.

Reply:
I'll think about more specific name, it is true that it may be a little bit too generic